### ISSUE TITLE
Filebeat does not recover when it fails to connect with k8s API

### ISSUE THREAD

**Issue by GreenKnight15** [elastic/beats#14164]
I am using the filebeat [elastic helm chart](https://github.com/elastic/helm-charts/tree/master/filebeat) under an [Istio service mesh](https://istio.io/docs/setup/getting-started/). Filebeat is configured to pull k8s meta data using the `add_kubernetes_metadata` processor. As of filebeat 7.4.0 with the new [k8s client #13630 (https://github.com/elastic/beats/pull/13051)  filebeat starts faster than the Istio side car which blocks outbound requests to the k8s API. After the failure filebeat never recovers the k8s connection and I lose all k8s meta data on my log packets from that point forward. I would expect the k8s client or filebeat to attempt to reestablish the connection and only a few number for logs would be missing their k8s meta data.

Error logs:
```
2019-10-17T20:33:29.733Z ERROR kubernetes/util.go:85 kubernetes: Querying for pod failed with error: Get https://10.100.0.1:443/api/v1/namespaces/bootstrap/pods/bootstrap-filebeat-x-pj8n9: dial tcp 10.100.0.1:443: connect: connection refused
E1017 20:33:29.734464 1 reflector.go:125] github.com/elastic/beats/libbeat/common/kubernetes/watcher.go:235: Failed to list *v1.Pod: Get https://10.100.0.1:443/api/v1/pods?fieldSelector=spec.nodeName%3Dlocalhost&limit=500&resourceVersion=0: dial tcp 10.100.0.1:443: connect: connection refused
```
I would share the istio logs for this outbound request but istio could never start before filebeat so the request was never captured.

Filebeat container configuration: 
```yaml
      - type: container
        enabled: true
        paths:
          - /var/lib/docker/containers/*/*.log
        stream: all
        processors:
        # This adds all of the kubernetes properties to the log packets i.e. pod_name, namespace, image, etc.
        - add_kubernetes_metadata: {}
        # This rename processor makes K8s pods using old label convention 'kubernetes.labels.app' to use the new convention 'kubernetes.labels.app.kubernetes.io/name'
        # This is necessary to avoid index pattern conflicts in Kibana which result in missing logs
        # If a pod ever has both the old and new conventions, 'app' will not be renamed and will appear in kibana as 'kubernetes.labels.app.value'
        - rename:
            when:
              and:
                - has_fields: ['kubernetes.labels.app']
                - not:
                    has_fields: ['kubernetes.labels.app.kubernetes.io/name']
            fields:
              - from: 'kubernetes.labels.app'
                to: 'kubernetes.labels.app_kubernetes_io/name'
            ignore_missing: true
            fail_on_error: false
```

For confirmed bugs, please report:
- Version: 7.4.0
- Operating System: Elastic Offical Docker image, Linux
  - docker.elastic.co/beats/filebeat:7.4.0

- Discuss Forum URL: https://discuss.elastic.co/t/filebeat-7-4-0-does-not-recover-when-it-fails-to-connect-with-k8s-api/204327
- Steps to Reproduce: 
Install the latest elastic filebeat helm chart into a k8s cluster with istio injection enabled. 
```
// Install isito: https://istio.io/docs/setup/install/helm/
// Inject namespace
kubectl label namespace default istio-injection=enabled
// Add elastic helm repo
helm repo add elastic https://helm.elastic.co
// Install filebeat
helm install --name filebeat elastic/filebeat --namespace default
```

Filebeat will start before the istio side car which causes the initial outbound k8s requests fail.
It is not an issue with the istio configuration it will still happen even if mTLS and RBAC is none restrictive.

**Comment by ephill** [comment]
Has anyone come up with a solid workaround for this? I can only come up with really hacky modifications to the official filebeat chart to prevent it from starting too early. Or, not using Istio on my filebeat pods. Neither feel like good options.

**Comment by GreenKnight15** [comment]
I have not found a solid solution, I'm still using a hacky mod to the filebeat container command to sleep before start up. It allows the side car to start in time. It's works most of the time but I'm not happy about it. Haven't heard anything from elastic :(

**Comment by rcny** [comment]
I've solved this with switching to the [autodiscover pattern](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-autodiscover.html). The key thing is to harvest what `kube-apiserver` has told you to harvest. Filebeat is able to filter pods by the node where Filebeat itself is scheduled.

**Comment by aeftef** [comment]
Same problem here, filebeat + istio.
I have deployed a workarround, overriding filebeat conatiner startup script so it first waits for the api server to be reachable
adding this parameter for filebeat container (I'm deploying it as a sidecar)

`command: ['/bin/bash', '-c','while [[ "$(/bin/curl -k -s -o /dev/null -w ''%{http_code}'' https://kubernetes.default.svc.cluster.local/version)" != "200" ]]; do sleep 2; done; /usr/local/bin/docker-entrypoint -e']
`

Note that the api server address is hardcoded and its ignoring certificate validation :( Improvements are welcome.

**Comment by FischlerA** [comment]
I just came across this problem as well.
The proper way to handle this issue would be for the filebeat to retry the connection until it is able to communikate with the api server.

Here is code from another issue mentioned as workaround for any application with this issue:
```
command: ["/bin/sh", "-c"]
args: ["until curl --head localhost:15000 ; do echo Waiting for Sidecar; sleep 1 ; done ; echo Sidecar available; echo Container 1 is Running ; startup.sh"] # replace startup.sh with actual startup command.
```

Since this issue is open since 2019; Is anybody from elastic on this?

**Comment by FischlerA** [comment]
This is still an issue that needs to be fixed...

**Comment by belimawr** [comment]
I'm trying to reproduce, but I'm not managing to. Filebeat is starting and working fine for me.

Here is what I did:
1. Started a Minikube cluster
2. Installed Helm
3. Installed Istio (following https://istio.io/latest/docs/setup/install/helm/), but I did not do the step 4 (optional)
4. Installed Filebeat 8.1 using the [current documentation](https://www.elastic.co/guide/en/beats/filebeat/current/running-on-kubernetes.html). It worked fine
5. Removed Filebeat 8.1
6. Installed Filebeat 7.17.1 using the helm charts from https://helm.elastic.co
7. Edited the configmap to add my credentials
8. Restarted Filebeat pod (I deleted it and let Kubernetes recreate it). It also worked fine.


@FischlerA , @GreenKnight15 are you still facing this problem? If so, could you help me to reproduce it?
